### ISSUE TITLE
Feature Request: Expanded Bulk Audit functionality

### ISSUE THREAD

**Issue by lostmydog** [snipe/snipe-it#4644]
#### Expected Behavior (or desired behavior if a feature request)
Currently, Bulk Audit is still in its infancy, I'd love for some expanded functionality to make the tool very useful for everyone. 

The workflow for a bulk audit should be:
First, select a location
Second, scan all assets in location
Once all assets are scanned, click audit button
Next window shows you a breakdown of items scanned. Showing for example: 5 assets audited, 3 assets correct, 2 assets scanned that were previously located elsewhere (Snipe-IT automatically updates the location of these 2 assets to the new room), 1 asset missing from scan (Snipe-IT prompts to assign the status to missing and location to unknown).

With this functionality, a user would be able to go room by room scanning their assets. As they are scanning, they are able to update the location of assets that were not correct, and mark the assets that are missing as such. If those missing items show up in another room, they will then be reassigned to an active status and the location updated for that room. I also believe that you should be able to reassign the user during the asset. I did not show it well in my image below, however I think a second field asking who each asset should be checked out to (or a drop down for the 'checked out to' column) in case the assets are in the correct location however the user has changed. 
![bulkaudit](https://user-images.githubusercontent.com/33837075/34123744-90ecbf74-e3fe-11e7-9af8-b0b691ae111c.png)

-----

#### Actual Behavior

(what actually happens goes here)

-----

#### Please confirm you have done the following before posting your bug report:

- [ ] I have enabled debug mode 
- [ ] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

-----
#### Provide answers to these questions:

- Is this a fresh install or an upgrade? 
- Version of Snipe-IT you're running
- Version of PHP you're running
- Version of MySQL/MariaDB you're running
- What OS and web server you're running Snipe-IT on
- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc)
- WITH DEBUG TURNED ON, if you're getting an error in your browser, include that error
- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error
- If a stacktrace is provided in the error, include that too.
- Any errors that appear in your browser's error console.
- Confirm whether the error is reproducible on the demo: https://snipeitapp.com/demo.
- Include any additional information you can find in `app/storage/logs` and your webserver's logs.
- Include what you've done so far in the installation, and if you got any error messages along the way.
- Indicate whether or not you've manually edited any data directly in the database

Please do not post an issue without answering the related questions above. If you have opened a different issue and already answered these questions, answer them again, once for every ticket. It will be next to impossible for us to help you.

https://snipe-it.readme.io/docs/getting-help

**Comment by bmanalili** [comment]
Can't wait for this be finished, very important update.

**Comment by lostmydog** [comment]
glad somebody else was looking for this. I thought there would be tons of comments on this one! This is going to be huge if/when it gets implemented. The ability to simply go room by room and audit is going to save us so much time.

**Comment by snipe** [comment]
While I’m okay with the proposed changes, I’m not sure how this is really different than the existing functionality. Scan asset, hit return, it gets added to the total audit session. Change the location if you need to. Keep scanning.

**Comment by lostmydog** [comment]
@snipe it looks like the existing functionality simply audits the assets existence, it does not update its location/owner. The functionality I'm looking for will allow users to say audit and update an entire room through the audit function. For example, i just audited a printer next to me that is currently 'ready to deploy' and its location is room 200.  I selected location '201' in the audit, however, going to that asset it still shows its location as room 200. In fact, the only place where I can see that the asset was located in room 201 is by going to the audit log and looking up the asset number. The functionality I'm looking for also adds the ability to update the owner, which is a huge addition in my line of work because there is significant turnover.

**Comment by snipe** [comment]
It will update the location if you select that in the audit screen, however it's not meant to update owners. Only checkout should update who actually has the device. Perhaps the ability to mark an audited asset as existing, but needing an owner update could be the best of both worlds, but we have no plans to add checkin/checkout to auditing, as that's not what they're for.

**Comment by lostmydog** [comment]
@snipe How is the location updated? I'm on v4.1.11, when I select a location and put in a tag number in bulk audit and select 'audit', the location is not updated in the assets profile

**Comment by bmanalili** [comment]
Adding to this,  it is also useful if the system will also tell you assets
that are found in the location but were not found during audit.  Indicating
item/s are lost-missing.

**Comment by PiQuer** [comment]
I am on version v4.4.1 - build 3636 (master), and I have the same issue: location is not updated during audit, selecting a location in the audit screen does not have any effect. @snipe could you please elaborate how this is supposed to work?
The other feature suggestions are also very useful, maybe this issue could be re-opened?

**Comment by mjb2000** [comment]
The OP mentioned scanning items in bulk, this would be great since it would allow an RFID reader to throw in a whole load of numbers to be audited (while barcode scanners will scan one-by-one an RFID reader will often pick up a bunch of tags at once, so it would be great to collect these either directly in to the web page, or copy them in from an RFID tag gathering app).

Will the bulk input of asset number be supported? Either CSV or CR/LF?

**Comment by snipe** [comment]
I'm sure it will be. I just picked up a NFC/RFID reader and some cards/stickers, so it really depends how my experiments go.

**Comment by jacobsen9026** [comment]
@snipe I love your system, it's so much easier to use than our old system Wasp Mobile Asset.

Only one thing that bothers me is this issue. It does not update the locations of assets with an existing location when they are not already checked out. 

I can reproduce this in the demo quite easily.

You mentioned

> It will update the location if you select that in the audit screen, however it's not meant to update owners. 

But it does not. Please let me know if you would like any more info but simply going to the demo and going to for example:

Asset 426236223

which has a location of Ashleighbury and is not already checked out

I choose to audit the asset

Select the location East Antionette

Hit audit and the location does not change.

Furthermore, the history on the asset shows an audit entry but no reference to the location so there is literally nowhere I can connect the dots.

Please advise.

**Comment by memtech3** [comment]
OH MY GOD THANK YOU SOOO MUCH

**Comment by memtech3** [comment]
Now snipe-it is starting to look a whole lot more like some of the paid asset management solutions

**Comment by gizzmojr** [comment]
In response to @jacobsen9026 comment about non-checked out assets getting location updates. I've recently need to do a 100% audit and I now need to print audit reports. My life would be so much easier if bulk scanning, with audit updates, were all aligned. 

I'd like to walk into a COMM room, cubicle, etc and just zap everything and hit update, regardless of what the current state of the asset is.

I've dusted off an old project that just needed a bit more work, which I'm very happily going to start using right away. Perhaps it might be of some use to you too.
https://github.com/gizzmojr/snipeit4-scanner

**Comment by snipe** [comment]
@gizzmojr i’m not clear in what you need changed here tho.

**Comment by memtech3** [comment]
do a pull request with your project

**Comment by vstephanou** [comment]
Sorry, I found the report. The rest of the comment remains.

**Comment by Turk186** [comment]
I would like to request the ability for self-audit. This would send out an email to the current asset checked out person people with check out assets to confirm asset location and person.  The screen should be customizable for what's displayed on the screen, like a list of assets, software, and options. Options could include things like, have asset listed, do not have an asset anymore with an option to type current asset they have or to select people they manage who have that asset. 

We have 40+ locations and a physical audit is done every 3 years, would like the ability to do a "soft" audit yearly.

**Comment by dnavre** [comment]
This functionality will help us a lot! 
Also, it can work slightly different. Our company is a bit bigger than 100 people and nearly all of them are sitting in an open space. So the room-by-room approach won't really work.
However, the audit can be done on a person by person basis. So an admin can go over the list of employees and re-affirm that a certain asset is still with the employee set in the database.

**Comment by memtech3** [comment]
YES!

**Comment by Turk186** [comment]
Yes. About to do an audit!

Norman Hibble
_________________________
"Knowledge is often mistaken for intelligence. This is like mistaking a cup of milk for a cow."

**Comment by Stacykg91** [comment]
Is this still not a thing? We've signed up to this recently and need to audit 11 schools.

**Comment by tritonaus** [comment]
Second requiring this feature.

We want to be able to rapidly update location by scanning assets in a store room/office to update their location. As others have mentioned staff move things around a lot, it would be great to just be able to go through, select a location and bulk update everything found to the location specified.

As mentioned previously, even when looking at the asset, it doesn't show it's location at audit, you have to go to the reports section and look up the asset number there.

Maybe just add a checkbox to update location on the audit page?

**Comment by memtech3** [comment]
Yes please!

**Comment by jacobsen9026** [comment]
This is the key missing feature for our use case.

**Comment by memtech3** [comment]
need to have a missing assets list per room

**Comment by memtech3** [comment]
highlight the location changed yellow or orange

**Comment by asmith0176** [comment]
Expanding on this feature would be huge. I do IT for our school district and audit around 4000 devices through the summer months. Many times we have new users coming and going so we need to quickly reassign these devices to these new users. I know in an earlier post it was said that Check-in/Checkout isn't planned to be implemented into the audit process but it would be an excellent feature. It would also be great to have the option to include custom fields into this. Maybe as you Audit device include a quick edit button, or even bulk edit on the list in case there are any other fields that need to modified for these devices quickly. 

Would it also be possible to lookup based on Serial Number rather than Asset tag? We use serial number to keep track of devices and not asset tags so non of our devices have Asset tags on them. Could we set those fields to search both Asset Tag and Serial Number or is there a way to allow that to be optional?

**Comment by timwsuqld** [comment]
This is still an issue. During an audit, locations are NOT updated to match the selected location

**Comment by jacobsen9026** [comment]
Can we get this issue reopened? It is still relevant.

**Comment by jacobsen9026** [comment]
What function is supposed to be handling this? Thinking of adding the functionality myself.

In AssetController.php under app/html/controllers I found two functions, the first appears to be the actual auditing function, the other takes a file as a parameter, I don't think that one is the one I want.


```
public function audit($id)
    {
        $settings = Setting::getSettings();
        $this->authorize('audit', Asset::class);
        $dt = Carbon::now()->addMonths($settings->audit_interval)->toDateString();
        $asset = Asset::findOrFail($id);
        return view('hardware/audit')->with('asset', $asset)->with('next_audit_date', $dt)->with('locations_list');
    }

```
Only issue is this function takes only an asset ID and no location. Gotta find what calls this and duplicate the function with a location parameter added.

There's another possible function I could use in the API folder of Controllers also called AssetController.php

```
public function audit(Request $request) {
        $this->authorize('audit', Asset::class);
        $rules = array(
            'asset_tag' => 'required',
            'location_id' => 'exists:locations,id|nullable|numeric',
            'next_audit_date' => 'date|nullable'
        );
        $validator = Validator::make($request->all(), $rules);
        if ($validator->fails()) {
            return response()->json(Helper::formatStandardApiResponse('error', null, $validator->errors()->all()));
        }
        $asset = Asset::where('asset_tag','=', $request->input('asset_tag'))->first();
        if ($asset) {
            // We don't want to log this as a normal update, so let's bypass that
            $asset->unsetEventDispatcher();
            $asset->next_audit_date = $request->input('next_audit_date');
            $asset->last_audit_date = date('Y-m-d h:i:s');
            if ($asset->save()) {
                $log = $asset->logAudit(request('note'),request('location_id'));
                return response()->json(Helper::formatStandardApiResponse('success', [
                    'asset_tag'=> e($asset->asset_tag),
                    'note'=> e($request->input('note')),
                    'next_audit_date' => Helper::getFormattedDateObject($log->calcNextAuditDate())
                ], trans('admin/hardware/message.audit.success')));
            }
        }
        return response()->json(Helper::formatStandardApiResponse('error', ['asset_tag'=> e($request->input('asset_tag'))], 'Asset with tag '.$request->input('asset_tag').' not found'));
    }

```
This one at least has a location_id, does some checks against it, then does nothing with it. This might be the one.

**Comment by jacobsen9026** [comment]
I have found a fix for the audit and update location. It is in the auditStore function in AssetsController.php in App/HTML/Controllers.

 Just gonna work on preventing it from affecting checked out assets, and then I'll post the changes needed.

**Comment by jacobsen9026** [comment]
I've posted my commits from my fork https://github.com/jacobsen9026/snipe-it


This is the updated auditStore function found in /App/HTML/Controllers/AssetsController.php

It checks if the asset is checked out, and if not, updates the location based on the input from the audit form.

   ```
public function auditStore(AssetFileRequest $request, $id)
    {
        $this->authorize('audit', Asset::class);

        $rules = array(
            'location_id' => 'exists:locations,id|nullable|numeric',
            'next_audit_date' => 'date|nullable'
        );

        $validator = \Validator::make($request->all(), $rules);

        if ($validator->fails()) {
            return response()->json(Helper::formatStandardApiResponse('error', null, $validator->errors()->all()));
        }

        $asset = Asset::findOrFail($id);

        // We don't want to log this as a normal update, so let's bypass that
        $asset->unsetEventDispatcher();

	//$asset->next_audit_date = date('Y-m-d h:i:s');
        $asset->next_audit_date = $request->input('next_audit_date');
        $asset->last_audit_date = date('Y-m-d h:i:s');

        if (is_null($target = $asset->assignedTo)){
        $asset->location_id = $request->input('location_id');
        }
        if ($asset->save()) {


            $filename = '';

            if ($request->hasFile('image')) {
                $file = $request->file('image');
                try {
                    $destinationPath = config('app.private_uploads').'/audits';
                    $extension = $file->getClientOriginalExtension();
                    $filename = 'audit-'.$asset->id.'-'.str_slug(basename($file->getClientOriginalName(), '.'.$extension)).'.'.$extension;
                    $file->move($destinationPath, $filename);
                } catch (\Exception $e) {
                    \Log::info($e);
                }
            }
            $location = Location::find(request('location_id'))['name'];
            $asset->logAudit($request->input('note').$location, $request->input('location_id'), $filename);
            return redirect()->to("hardware")->with('success', trans('admin/hardware/message.audit.success'));
        }
    }

```
But that doesn't handle the bulk audit. For that we need to update the audit function in /App/HTML/Controllers/API/AssetsController.php to be the following. Does the same stuff as the other file.

 ```
   public function audit(Request $request) {


        $this->authorize('audit', Asset::class);
        $rules = array(
            'asset_tag' => 'required',
            'location_id' => 'exists:locations,id|nullable|numeric',
            'next_audit_date' => 'date|nullable'
        );

        $validator = Validator::make($request->all(), $rules);
        if ($validator->fails()) {
            return response()->json(Helper::formatStandardApiResponse('error', null, $validator->errors()->all()));
        }

        $asset = Asset::where('asset_tag','=', $request->input('asset_tag'))->first();


        if ($asset) {
            // We don't want to log this as a normal update, so let's bypass that
            $asset->unsetEventDispatcher();

            $asset->next_audit_date = $request->input('next_audit_date');
            $asset->last_audit_date = date('Y-m-d h:i:s');
            if (is_null($target = $asset->assignedTo)){
               $asset->location_id = $request->input('location_id');
            }


            if ($asset->save()) {
                $location = Location::find(request('location_id'))['name'];
                $log = $asset->logAudit(request('note').$location,request('note'));

                return response()->json(Helper::formatStandardApiResponse('success', [
                    'asset_tag'=> e($asset->asset_tag),
                    'note'=> e($request->input('note')),
                    'next_audit_date' => Helper::getFormattedDateObject($log->calcNextAuditDate())
                ], trans('admin/hardware/message.audit.success')));
            }
        }

        return response()->json(Helper::formatStandardApiResponse('error', ['asset_tag'=> e($request->input('asset_tag'))], 'Asset with tag '.$request->input('asset_tag').' not found'));





    }

```

**Comment by jacobsen9026** [comment]
I also saw last night that this functionality is already being built into the Development branch. (I accidentally committed to the Master branch at first).

**Comment by jacobsen9026** [comment]
For those interested in moving items that are checked out to location with an audit replace the statement block 
```
if (is_null($target = $asset->assignedTo)){
              $asset->location_id = $request->input('location_id');
           }
```
with
```
if ($asset->assignedType() == Asset::LOCATION){
                $target = Location::find($request->input('location_id'));
                $asset->checkOut($target, Auth::user(), date('Y-m-d H:i:s'), '', 'Checked out on asset audit');
            }

```
In both AssetsController.php files. This is how I'm currently running.

**Comment by jacobsen9026** [comment]
Pull request 
https://github.com/snipe/snipe-it/pull/7580

I added an additional setting in the General Settings section to always checkout audited assets if they are not already checked out to a user.

**Comment by rmacaldo** [comment]
Is this still available? I'm on v6.0.14 build 9236 but nothing seems to appear on the feature request.

**Comment by jacobsen9026** [comment]
It doesn't appear that my pull request was accepted. They still have a broken audit, even the notes section to this day is still useless as it's data is inaccessible from the website, if there is any.

I've been running my customized version of this app for my job same as the pull request I submitted, for a while. Definitely missing some security updates by now, but I keep it off the internet so not too much of a risk.
